ncron.1
.\" Man page for ncron
.\"
.\" Copyright (c) 2004-2026 Nicholas J. Kain
.\"
.TH ncron 1 "July 14, 2026"
.LO 1
.SH NAME
ncron \- secure and reliable cron/at daemon
.SH SYNOPSIS
ncron [\-b0jqhv] [\-c config_file] [\-t crontab_file]
      [\-H history_file]
.SH DESCRIPTION
.B ncron
runs programs at intervals specified by the user, subject to time constraints.
It is designed to be both secure and reliable.
.SH OPTIONS
.TP
.B \-\^c , \-\-config=FILE
Specify the configuration file to use.  Configuration file elements are
specified as key = value pairs, and correspond to long command line arguments.
.TP
.B \-\^t , \-\-crontab=FILE
Specify the crontab file to use; the default is
.BR /var/lib/ncron/crontab .
(See
.BR crontab (5).)
.TP
.B \-\^H , \-\-history=FILE
Specify the file in which ncron will store job
runtimes and the number of times that a job has
been invoked.  The default location is
.BR /var/lib/ncron/history .
.TP
.B \-\^0   \-\-noexecsave
Do not save any data on the times when jobs are executed.
.TP
.B \-\^j   \-\-journal
Save execution times to disk immediately after a job is run.  Perhaps useful
for very unstable machines that often suffer unclean shutdowns.
.TP
.BR \-\^h  \-\-help
Print abbreviated help and exit.
.TP
.B \-\^v   \-\-version
Print version information and exit.
.TP

.SH "INTRODUCTION"
.PP
ncron's job scheduling has two critical concepts: intervals and constraints. An
"interval" is simply the minimum time between job invocations. The concept of
time intervals is no doubt familiar. A trivial example of a simple interval
would be "wait one day before synchronizing my clock again". ncron intervals
are specified in seconds and have second precision.
.PP
Constraints limit the times in which a particular job may be run.  In
ncron, intervals may be constrained by ranges of minutes, hours,
weekdays, days of the month, and months of the year.  Multiple
constraint ranges may be applied, even of the same type.  Constraints
of the same type will be mathematically unioned (logical OR'ed)
together. ncron internally uses a sieve-based algorithm to find the
smallest suitable time that satisfies the specified interval within
the specified constraint.
.PP
Examples are probably the clearest way of explaining this system.  Suppose we
would like a locatedb to run weekly between 1 and 5am.  Since we want the
program to run weekly, we would specify an interval of interval=7w.  We would
then specify a constraint of time=1:00-5:00.  ncron would then make sure that our job
only ran weekly between 1 and 5am.
.PP
The astute reader might wonder why ncron deals with constraints in terms of
intervals -- after all, if ncron uses a greedy algorithm, the task will only
run at a time other than the lower bound of the interval if the cron is invoked
at a time after the lower bound of the interval but within the interval. Why
bother with unnecessary complexity to handle that case? Traditional cronds
certainly don't bother. The answer lies in ncron's attempts to be reliable and
fast.
.PP
ncron internally works by precalculating the exact time at which a job
should be run. ncron then spends most of its time sleeping until its
next job, waking up only to handle signals. If ncron is gracefully
terminated, it will by default save these times it has
calculated. Thus, ncron keeps a memory of when jobs were invoked in
the past. This feature is obviously quite useful for systems that are
sometimes either suspended or shut down.  For those who have very
critical jobs or very unstable computers, ncron can optionally save
these times to disk whenever they change in memory. This mode of
operation is called "journalling".  Journal mode may be set either for
all jobs with a command-line flag, or for particular jobs in the
crontab file.
.PP
If it is necessary for regular users to run periodic background tasks, the
recommended method is to have the users run a per-user ncron.  This is easily
possible by using the -c and -H options to specify user-specific crontab
and history files.  Tasks will then naturally run as that user.
.PP
For specifics on how to configure ncron, consult
.B crontab(5).

.SH ENVIRONMENT
ncron ignores its environment. It will pass through the environment which it
was provided when it was invoked. If a job is to be run with a custom shell
environment, a wrapper script should be used.

.SH "SEE ALSO"
crontab(5).
.SH BUGS
None known, otherwise they would be fixed.
.SH AUTHOR
Nicholas J. Kain <njkain (at) gmail.com>



ChangeLog
user:        njk@darksword.kain.us
date:        Mon Sep 22 22:53:37 2008 -0400
summary:     Make makefile more sane.

user:        njk@darksword.kain.us
date:        Mon Sep 22 22:33:31 2008 -0400
summary:     Convert unnecessary static char arrays to char arrays.  Only one is needed.

user:        njk@darksword.kain.us
date:        Mon Sep 22 22:17:01 2008 -0400
summary:     cosmetic: Cleaned up code formatting.

user:        njk@darksword.kain.us
date:        Thu Feb 08 05:59:40 2007 -0500
summary:     Add w=week interval unit.  Update documentation to remove duplication.

user:        njk@darksword.kain.us
date:        Thu Feb 08 05:40:14 2007 -0500
summary:     Clean up Makefile and whitespace/identing

user:        njk@darksword.kain.us
date:        Thu Feb 08 05:11:32 2007 -0500
summary:     Make parse_interval() static

user:        njk@darksword.kain.us
date:        Thu Feb 08 03:38:03 2007 -0500
summary:     Allow unit specifiers for intervals and update documentation.

user:        njk@darksword.kain.us
date:        Thu Feb 08 02:50:43 2007 -0500
summary:     Fix bug that broke single value rlimit pairs

user:        njk@darksword.kain.us
date:        Thu Feb 08 02:49:08 2007 -0500
summary:     Fix bug that broke single-item ipairs

user:        njk@darksword.kain.us
date:        Thu Feb 08 02:45:53 2007 -0500
summary:     Initial commit

2004-09-06:
use clearenv if HAVE_CLEARENV is defined.  default for glibc compiles.

2004-07-16:
it's unnecessary to conditionalize free on a structure that does
not itself contain pointers.  the implied changes were performed on sched.c
and config.c.

ncron now uses darcs for revision control.

2004-06-26:
signal flag variables are now static volatile sig_atomic_t rather than
static volatile int.

2004-03-06:

ncron is now intelligent enough to automatically compensate for constraint
changes for extant jobs.  This change means that users need not clear
exectime data in ncron's history file for edited jobs.

2004-03-05:

"-Wall -pedantic" are now default flags in builds.
Minor code cleanups to satisfy "-Wall -pedantic".

ncron now has friendly man pages.

version is now 0.98

2004-03-04:

Constrainer is now more intelligent about selecting intervals, and should
always pick the most optimal runtime.

Constrainer now understands that the number of days in a month varies as a
function of the month and year.

Constrainer should be faster and generally more correct.

Config parser now drops a job onto the dead list if initial constraint fails.

2004-03-02:

Minor code cleanups.

Add "runat" and "journal" keywords.

Improved ncron so that exectime and number of runs for expired jobs will
be saved.  However, this comes at a cost -- ncron will keep data structures
for jobs that have expired in memory.  This problem may be fixed by either
dividing the current history file into two files (preferred) or alternatively,
by having ncron edit the user configuration file to delete expired jobs (bad!).
For now, I don't consider it to be a big issue.

Make nstrl.c idempotent.

Added "maxruns" keyword.  In order to support this functionality properly,
number of job runs is now recorded along with execution times.

2004-02-26:

Initial constraint for jobs without prior exectime data does not first
add interval before constraining.

2004-02-25:

Multiple constraints of the same type are now allowed; they are implicitly
OR'ed together.

Discontinuous constraints should be more reliable.
Lots of internal improvements.

waitpid with WNOHANG never returns EINTR.

Makefile fixes

I have "ported" ncron so that it can be cleanly built on OS X.  I have no idea
whether or not it will actually work properly, since I do not own any OS X
machines.

ncron now uses the daemon() syscall rather than homebrew fork code.  This
change may introduce portability issues.

Version is now 0.95.

2003-09-27:

No more repository -- paranoid security fears about the existing ones.
(In reality, my local repo was claimed by a deathstar crash.)

Did an audit for fun:
Minor code cleanup in config.c.
Unchecked malloc() in exec.c fixed.

Added native support to chroot spawned jobs.  This change is only useful
if ncron is being run uid root.



ncron.cpp
// Copyright 2003-2024 Nicholas J. Kain <njkain at gmail dot com>
// SPDX-License-Identifier: MIT
#include <unistd.h>
#include <stdio.h>
#include <string.h>

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <time.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <signal.h>
#include <errno.h>
#include <limits.h>
#include <getopt.h>

#ifdef __linux__
#include <sys/prctl.h>
#endif

extern "C" {
#include "nk/log.h"
#include "nk/io.h"
#include "xmalloc.h"
#include "strconv.h"
}
#include "ncron.hpp"
#include "sched.hpp"

#define CONFIG_FILE_DEFAULT "/var/lib/ncron/crontab"
#define EXEC_FILE_DEFAULT "/var/lib/ncron/exectimes"

#define NCRON_VERSION "3.0"

int gflags_debug;
static volatile sig_atomic_t pending_save_and_exit;
static int s6_notify_fd = -1;

/* Time (in msec) to sleep before dispatching events at startup.
   Set to a nonzero value so as not to compete for cpu with init scripts at
   boot time. */
static unsigned g_initial_sleep = 0;

static char const *g_ncron_conf = CONFIG_FILE_DEFAULT;
static char const *g_ncron_execfile = EXEC_FILE_DEFAULT;
static char const *g_ncron_execfile_tmp = EXEC_FILE_DEFAULT "~";
enum Execmode
{
    Execmode_normal = 0,
    Execmode_journal,
    Execmode_nosave,
};
static Execmode g_ncron_execmode = Execmode_normal;

size_t g_njobs;
Job *g_jobs;
static Job * stackl;
static Job * deadstackl;

static bool do_save_stack(FILE *f, Job *j)
{
    for (; j; j = j->next_) {
        if (fprintf(f, "%d=%li:%u|%lu\n", j->id_, j->exectime_, j->numruns_, j->lasttime_) < 0) {
            log_line("Failed to write to history file %s", g_ncron_execfile_tmp);
            return false;
        }
    }
    return true;
}

[[nodiscard]] static bool save_stack()
{
    FILE *f = fopen(g_ncron_execfile_tmp, "w");
    if (!f) {
        log_line("Failed to open history file %s for write", g_ncron_execfile_tmp);
        return false;
    }
    if (!do_save_stack(f, stackl)) goto err1;
    if (!do_save_stack(f, deadstackl)) goto err1;
    fclose(f);

    if (rename(g_ncron_execfile_tmp, g_ncron_execfile)) {
        log_line("Failed to update history file (%s => %s): %s",
                 g_ncron_execfile_tmp, g_ncron_execfile, strerror(errno));
        goto err0;
    }
    return true;
err1:
    fclose(f);
err0:
    unlink(g_ncron_execfile_tmp);
    return false;
}

static void save_and_exit(void)
{
    if (g_ncron_execmode != Execmode_nosave) {
        if (save_stack()) {
            log_line("Saved stack to %s.", g_ncron_execfile);
        } else {
            log_line("Failed to save stack to %s; some jobs may run again.",
                     g_ncron_execfile);
        }
    }
    // Get rid of leak sanitizer noise.
    for (size_t i = 0; i < g_njobs; ++i) job_destroy(&g_jobs[i]);
    log_line("Exited.");
    exit(EXIT_SUCCESS);
}

static void signal_handler(int sig)
{
    int serrno = errno;
    if (sig == SIGTERM || sig == SIGINT || sig == SIGHUP) {
        pending_save_and_exit = 1;
    }
    errno = serrno;
}

static void fix_signals(void)
{
    static const int ss[] = {
        SIGHUP, SIGINT, SIGTERM, SIGKILL
    };
    sigset_t mask;
    if (sigprocmask(0, 0, &mask) < 0)
        suicide("sigprocmask failed");
    for (int i = 0; ss[i] != SIGKILL; ++i)
        if (sigdelset(&mask, ss[i]))
            suicide("sigdelset failed");
    if (sigaddset(&mask, SIGPIPE))
        suicide("sigaddset failed");
    if (sigprocmask(SIG_SETMASK, &mask, nullptr) < 0)
        suicide("sigprocmask failed");

    struct sigaction sa;
    memset(&sa, 0, sizeof sa);
    sa.sa_handler = signal_handler;
    sa.sa_flags = SA_RESTART;
    if (sigemptyset(&sa.sa_mask))
        suicide("sigemptyset failed");
    for (int i = 0; ss[i] != SIGKILL; ++i)
        if (sigaction(ss[i], &sa, NULL))
            suicide("sigaction failed");
    sa.sa_handler = SIG_IGN;
    sa.sa_flags = SA_NOCLDWAIT;
    if (sigaction(SIGCHLD, &sa, NULL))
        suicide("sigaction failed");
}

static void fail_on_fdne(char const *file, int mode)
{
    if (access(file, mode)) {
        log_line("File '%s' does not exist or is not %s",
                 file, (mode & W_OK) ? "writable" : "readable");
        exit(EXIT_FAILURE);
    }
}

static void sleep_or_die(struct timespec *ts)
{
retry:
    int r = clock_nanosleep(CLOCK_REALTIME, TIMER_ABSTIME, ts, NULL);
    if (r) {
        if (r == EINTR) {
            if (pending_save_and_exit) save_and_exit();
            goto retry;
        }
        log_line("clock_nanosleep failed: %s", strerror(r));
        exit(EXIT_FAILURE);
    }
}

void clock_or_die(struct timespec *ts)
{
    if (clock_gettime(CLOCK_REALTIME, ts)) {
        log_line("clock_gettime failed: %s", strerror(errno));
        exit(EXIT_FAILURE);
    }
}

static inline void debug_stack_print(const struct timespec &ts) {
    if (!gflags_debug)
        return;
    if (stackl)
        log_line("ts.tv_sec = %lu  stack.front().exectime = %lu", ts.tv_sec, stackl->exectime_);
    for (Job *j = stackl; j; j = j->next_)
        log_line("job %d exectime = %lu", j->id_, j->exectime_);
}

static void do_work(unsigned initial_sleep)
{
    struct timespec ts;
    clock_or_die(&ts);
    ts.tv_sec += initial_sleep / 1000;
    ts.tv_nsec += (initial_sleep % 1000) * 1000000;

    bool pending_save = false;
    for (;;) {
        if (pending_save) {
            if (!save_stack()) {
                log_line("Failed to save stack to %s for a journalled job.",
                         g_ncron_execfile);
            } else {
                pending_save = false;
            }
        }
        sleep_or_die(&ts);

        while (stackl->exectime_ <= ts.tv_sec) {
            Job *j = stackl;
            if (gflags_debug)
                log_line("DISPATCH %d (%lu <= %lu)", j->id_, j->exectime_, ts.tv_sec);

            job_exec(j, &ts);
            if (j->journal_ || g_ncron_execmode == Execmode_journal)
                pending_save = true;

            if ((j->numruns_ < j->maxruns_ || j->maxruns_ == 0) && j->exectime_ != 0) {
                if (stackl->next_) {
                    Job *t = stackl;
                    stackl = stackl->next_;
                    job_insert(&stackl, t);
                }
            } else {
                Job *t = stackl;
                stackl = stackl->next_;
                job_insert(&deadstackl, t);
            }
            if (!stackl)
                save_and_exit();
        }

        debug_stack_print(ts);
        {
            Job *j = stackl;
            if (ts.tv_sec <= j->exectime_) {
                time_t tdelta = j->exectime_ - ts.tv_sec;
                ts.tv_sec = j->exectime_;
                ts.tv_nsec = 0;
                if (gflags_debug)
                    log_line("SLEEP %zu seconds", tdelta);
            }
        }
    }
}

static void usage()
{
    printf("ncron " NCRON_VERSION ", cron/at daemon.\n"
           "Copyright 2003-2024 Nicholas J. Kain\n"
           "Usage: ncron [options]...\n\nOptions:\n"
           "--help         -h    Print usage and exit.\n"
           "--version      -v    Print version and exit.\n"
           "--sleep        -s [] Initial sleep time in seconds.\n"
           "--noexecsave   -0    Don't save execution history at all.\n"
           "--journal      -j    Save exectimes at each job invocation.\n"
           "--crontab      -t [] Path to crontab file.\n"
           "--history      -H [] Path to execution history file.\n"
           "--verbose      -V    Log diagnostic information.\n"
    );
}

static char *xstrdup(const char *s)
{
    char *r = strdup(s);
    if (!r) exit(EXIT_FAILURE);
    return r;
}

static void print_version()
{
    log_line("ncron " NCRON_VERSION ", cron/at daemon.\n"
             "Copyright 2003-2024 Nicholas J. Kain\n\n"
"Permission is hereby granted, free of charge, to any person obtaining\n"
"a copy of this software and associated documentation files (the\n"
"\"Software\"), to deal in the Software without restriction, including\n"
"without limitation the rights to use, copy, modify, merge, publish,\n"
"distribute, sublicense, and/or sell copies of the Software, and to\n"
"permit persons to whom the Software is furnished to do so, subject to\n"
"the following conditions:\n\n"
"The above copyright notice and this permission notice shall be\n"
"included in all copies or substantial portions of the Software.\n\n"
"THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND,\n"
"EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF\n"
"MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND\n"
"NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE\n"
"LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION\n"
"OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION\n"
"WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
             );
}

static void process_options(int ac, char *av[])
{
    static struct option long_options[] = {
        {"help", 0, nullptr, 'h'},
        {"version", 0, nullptr, 'v'},
        {"sleep", 1, nullptr, 's'},
        {"noexecsave", 0, nullptr, '0'},
        {"journal", 0, nullptr, 'j'},
        {"crontab", 1, nullptr, 't'},
        {"history", 1, nullptr, 'H'},
        {"s6-notify", 1, nullptr, 'd'},
        {"verbose", 0, nullptr, 'V'},
        {nullptr, 0, nullptr, 0 }
    };
    for (;;) {
        int c = getopt_long(ac, av, "hvbs:0jt:H:d:V", long_options, nullptr);
        if (c == -1) break;
        switch (c) {
            case 'h': usage(); exit(EXIT_SUCCESS); break;
            case 'v': print_version(); exit(EXIT_SUCCESS); break;
            case 's': if (!strconv_to_u32(optarg, optarg + strlen(optarg), &g_initial_sleep)) {
                          log_line("invalid sleep '%s' specified", optarg);
                          exit(EXIT_FAILURE);
                      }
                      break;
            case '0': g_ncron_execmode = Execmode_nosave; break;
            case 'j': g_ncron_execmode = Execmode_journal; break;
            case 't': g_ncron_conf = xstrdup(optarg); break;
            case 'H': {
                size_t l = strlen(optarg);
                g_ncron_execfile = xstrdup(optarg);
                char *tmpf = static_cast<char *>(xmalloc(l + 2));
                memcpy(tmpf, optarg, l);
                tmpf[l] = '~';
                tmpf[l+1] = 0;
                g_ncron_execfile_tmp = tmpf;
                break;
            }
            case 'd': s6_notify_fd = atoi(optarg); break;
            case 'V': gflags_debug = 1; break;
            default: break;
        }
    }
}

int main(int argc, char* argv[])
{
    process_options(argc, argv);
    fail_on_fdne(g_ncron_conf, R_OK);
    fail_on_fdne(g_ncron_execfile, R_OK | W_OK);
    parse_config(g_ncron_conf, g_ncron_execfile, &stackl, &deadstackl);

    if (!stackl) {
        log_line("No jobs, exiting.");
        exit(EXIT_FAILURE);
    }

    umask(077);
    fix_signals();

#ifdef __linux__
    prctl(PR_SET_DUMPABLE, 0, 0, 0, 0);
    prctl(PR_SET_KEEPCAPS, 0, 0, 0, 0);
    prctl(PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0);
#endif

    if (s6_notify_fd >= 0) {
        char buf[] = "\n";
        safe_write(s6_notify_fd, buf, 1);
        close(s6_notify_fd);
    }

    do_work(g_initial_sleep);
    exit(EXIT_SUCCESS);
}

